packages/expo-router/src/global-state/routing.ts

```diff
@@ -145,7 +145,8 @@ function getNavigateAction(state: ResultState, parentState: NavigationState, typ
   const routesAreEqual = parentState.routes[parentState.index] === currentRoute;
 
   // If there is nested state and the routes are equal, we should keep going down the tree
-  if (route.state && routesAreEqual && currentRoute.state) {
+  // unless the type is push, in which case we should just push the state
+  if (route.state && routesAreEqual && currentRoute.state && type !== 'PUSH') {
```

Comment:
`type !== 'PUSH'` seems wrong as we have this guard just below

https://github.com/expo/expo/blob/b5ad7c606700c0bff217ee036bf032c93adcc1c6/packages/expo-router/src/global-state/routing.ts#L158

Reply:
could be that it's because I'm using a drawer and not tabs?

Reply:
getting rid of this seems to fix my issue as well. I'm not sure why it needs to change the type if on the function call it defaults to NAVIGATE anyways.
https://github.com/expo/expo/blob/b5ad7c606700c0bff217ee036bf032c93adcc1c6/packages/expo-router/src/global-state/routing.ts#L156-L157

Reply:
never mind, that's not entirely true.

Reply:
It has something to do with routesAreEqual. It's comparing matching states when the route in the new state doesn't match. The only thing changing is the params. In the screen shot the sort param changed but the route doesn't push.
![image](https://github.com/expo/expo/assets/53270830/0b333ff6-31e6-4466-a950-2f789bc9b8e8)

Reply:
stepping away from this before I get homicidal

Reply:
Ok, so I think I found the pain points

First is routesAreEqual
https://github.com/expo/expo/blob/b5ad7c606700c0bff217ee036bf032c93adcc1c6/packages/expo-router/src/global-state/routing.ts#L145

routesAreEqual is returning true even though they are not the same. This is because objects are compared by reference, not by value. When comparing objects using == or ===, JavaScript checks to see if the objects reference the same location in memory, not if their contents are the same. Which in this case matters because the params have changed.

Second is currentRoute
https://github.com/expo/expo/blob/b5ad7c606700c0bff217ee036bf032c93adcc1c6/packages/expo-router/src/global-state/routing.ts#L144

This returns the first route in the list that matches the name, so when doing the comparison, it's comparing the first route rather than the last route (current route)

I think I'm close to a working solution with the functionality intended which is not to push the route if it is exactly the same.